### ISSUE TITLE
IO:  unhandled exception, in case the URL cannot be opened

### ISSUE THREAD

**Issue by kelvinwop** [scikit-image/scikit-image#3785]
## Description
![image](https://user-images.githubusercontent.com/30770221/53887634-4a1eb600-3fe8-11e9-97db-f374815eb620.png)

## Way to reproduce
```python
def url_to_image(url):
    """download image from url and return it"""
    im_mat = io.imread(url)
    image = cv2.imdecode(im_mat, cv2.IMREAD_COLOR)
 
    # return the image
    return image
```

## Version information
```python
>>> from __future__ import print_function
>>> import sys; print(sys.version)
3.7.2 (tags/v3.7.2:9a3ffc0492, Dec 23 2018, 23:09:28) [MSC v.1916 64 bit (AMD64)]
>>> import platform; print(platform.platform())
Windows-10-10.0.17134-SP0
>>> import skimage; print("scikit-image version: {}".format(skimage.__version__))
scikit-image version: 0.14.2
>>> import numpy; print("numpy version: {}".format(numpy.__version__))
```

```python
Traceback (most recent call last):
  File "C:\Python37\lib\site-packages\skimage\io\util.py", line 28, in file_or_url_context
    with tempfile.NamedTemporaryFile(delete=False, suffix=ext) as f:
  File "C:\Python37\lib\tempfile.py", line 547, in NamedTemporaryFile
    (fd, name) = _mkstemp_inner(dir, prefix, suffix, flags, output_type)
  File "C:\Python37\lib\tempfile.py", line 258, in _mkstemp_inner
    fd = _os.open(file, flags, 0o600)
OSError: [Errno 22] Invalid argument: 'C:\\Users\\kelvin\\AppData\\Local\\Temp\\tmpegjt2y8r.png?mi_guid=CFCDF62B5115D31389F2893D3CCFFA5CB42BDE7F8CD696BB57B97B8BB567940F&mi_language=EN'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "imagecheck.py", line 28, in <module>
    print(url_to_image("http://www.movable-ink-6437.com/p/rp/0fd853751b1edbb0.png?mi_guid=CFCDF62B5115D31389F2893D3CCFFA5CB42BDE7F8CD696BB57B97B8BB567940F&mi_language=EN"))
  File "imagecheck.py", line 9, in url_to_image
    im_mat = io.imread(url)
  File "C:\Python37\lib\site-packages\skimage\io\_io.py", line 61, in imread
    with file_or_url_context(fname) as fname:
  File "C:\Python37\lib\contextlib.py", line 112, in __enter__
    return next(self.gen)
  File "C:\Python37\lib\site-packages\skimage\io\util.py", line 34, in file_or_url_context
    os.remove(f.name)
UnboundLocalError: local variable 'f' referenced before assignment
```

**Comment by stefanv** [comment]
Sloppy programming there—it's an unhandled exception, in case the URL cannot be opened.  I tried it now, and it works, so I guess it must have been an intermittent network error.

Either way, I can verify that the bug exists and should be fixed by catching any error occurring while downloading the data.

**Comment by kelvinwop** [comment]
For the future generations:

The easier way to get an image for cv2 from a url is using imutils' function

```python
imutils.url_to_image(url)
```

**Comment by hmaarrfk** [comment]
you can also try `imageio`

**Comment by hmaarrfk** [comment]
I guess I can't reproduce easily. @rfezzani is trying to fix this, but I'm not sure the fix is the "correct" one.

**Comment by hmaarrfk** [comment]
I think the origininal error has been fixed in an other issue, where we stripped the `?asdlfakjsdlkafjlasdkfjaldskjlfak` at the end of the URL.

**Comment by hmaarrfk** [comment]
Was this already fixed in https://github.com/scikit-image/scikit-image/pull/3642

**Comment by stefanv** [comment]
Looking at the code, this doesn't seem right:

```python
        try:
            with tempfile.NamedTemporaryFile(delete=False, suffix=ext) as f:
                u = urllib.request.urlopen(resource_name)
                f.write(u.read())
            # f must be closed before yielding
            yield f.name
        finally:
            os.remove(f.name)
```

What happens if `NamedTemporaryFile` fails?

**Comment by rfezzani** [comment]
I observed in #3843 that `NamedTemporaryFile` raises

- an `urllib.error.URLError` if the service is unknown
- an `urllib.error.HTTPError` in case of HTTP errors.

#3843 attempts to fix #3785.

**Comment by rfezzani** [comment]
Looking at `NamedTemporaryFile` [source code](https://github.com/python/cpython/blob/2.7/Lib/tempfile.py#L447), the following errors may be raised:

- `FileNotFoundError` if no usable temporary directory is found ([link](https://github.com/python/cpython/blob/3.7/Lib/tempfile.py#L227))
- `FileExistsError` if no usable temporary file name found ([link](https://github.com/python/cpython/blob/3.7/Lib/tempfile.py#L271))
- `PermissionError` if temporary files can't be created ([link](https://github.com/python/cpython/blob/3.7/Lib/tempfile.py#L268))
- `BaseExeption` if `io.open` fails ([link](https://github.com/python/cpython/blob/3.7/Lib/tempfile.py#L556))

**Comment by hmaarrfk** [comment]
There are so many bad things that happen to your computer whne you can't create files. 
skimage will be the least of your worries. 

I doubt it is worth our time fixing something like this is worth our time unless we get a specific usecase of somebody being affected.

**Comment by jni** [comment]
@hmaarrfk 

> I doubt it is worth our time fixing something like this is worth our time unless we get a specific usecase of somebody being affected.

the original poster was affected! UnboundLocalErrors should never happen in scikit-image, as it is a symptom of sloppy code. It's only one step above SyntaxErrors. Are we happy to leave those around now as long as no one notices???

> There are so many bad things that happen to your computer whne you can't create files.
skimage will be the least of your worries.

Ironically this happened to me a couple of days ago. For some reason Ubuntu started seeing my fs as read-only. Even bash completion failed! :man_facepalming: 

Nevertheless, see above. ;) scikit-image should raise meaningful errors in such situations, e.g. if running from a LiveCD or whatever.

Probably/hopefully addressed in #4172.
